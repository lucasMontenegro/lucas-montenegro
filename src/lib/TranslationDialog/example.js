import { makeStyles } from "@material-ui/core/styles"
import React from "react"
import languageDetector from "lib/languageDetector"
import AppBar from "@material-ui/core/AppBar"
import Toolbar from "@material-ui/core/Toolbar"
import TranslationDialog from "lib/TranslationDialog"
import { ThemeProvider } from "@material-ui/styles"
import themes from "lib/paperbase"
import { Route } from "react-router-dom"
const languageCodes = [`en`, `es`]
const otherProps = {
  en: {
    key: `en`,
    id: `translate-to-en`,
    text: `English`,
  },
  es: {
    key: `es`,
    id: `translate-to-es`,
    text: `Spanish`,
  },
}
const links = [`dark`, `light`].reduce((links, themeName) => {
  links[themeName] = languageCodes.reduce((byLanguage, oldLanguage) => {
    byLanguage[oldLanguage] = languageCodes.map(newLanguage => ({
      languageCode: newLanguage,
      location: (
        newLanguage === oldLanguage ? null :
        { pathname: `/TranslationDialog/${themeName}/${newLanguage}` }
      ),
      otherProps: otherProps[newLanguage],
    }))
    return byLanguage
  }, {})
  return links
}, {})
const useStyles = makeStyles(theme => ({
  root: {
    minHeight: `100vh`,
    backgroundColor: theme.palette.background.default,
  },
}))
let count = 0
function Example ({ themeName }) {
  function getLinks () {
    console.log(`getLinks ${count}`)
    count++
    return links[themeName][languageDetector.get()]
  }
  const classes = useStyles()
  return (
    <div className={classes.root}>
      <AppBar><Toolbar><TranslationDialog getLinks={getLinks}/></Toolbar></AppBar>
    </div>
  )
}
function Init (props) {
  languageDetector.init(languageCodes)
  const ready = languageDetector.useReadyState()
  if (ready) {
    const { themeName, currentLanguage } = props.match.params
    languageDetector.set(currentLanguage)
    return (
      <ThemeProvider theme={themes[themeName].body || themes.light.body}>
        <Example themeName={themeName} />
      </ThemeProvider>
    )
  }
  return null
}
export default (
  <Route exact path="/TranslationDialog/:themeName/:currentLanguage" component={Init} />
)