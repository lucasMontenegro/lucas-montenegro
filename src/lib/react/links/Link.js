import React from "react"
import { Link as RouterLink } from "react-router-dom"
import MuiLink from "@material-ui/core/Link"
const ForwardLinkRef = React.forwardRef(function ForwardLinkRef (props, ref) {
  return <RouterLink {...props} innerRef={ref} />
})
export default function Link (props) {
  return <MuiLink {...props} component={ForwardLinkRef} />
}