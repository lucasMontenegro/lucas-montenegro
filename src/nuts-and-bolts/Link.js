import React from 'react';
import { Link as RouterLink } from 'react-router-dom';
import MaterialLink from '@material-ui/core/Link';
import clone from 'lodash/clone';

class Link extends React.Component {
  Link = React.forwardRef((itemProps, ref) => (
    <RouterLink {...itemProps} to={this.props.to} innerRef={ref} />
  ));

  render() {
    const props = clone(this.props);
    props.component = this.Link;
    delete props.to;
    return <MaterialLink {...props} />;
  }
}

export default Link;
