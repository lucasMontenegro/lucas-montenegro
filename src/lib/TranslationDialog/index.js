import Translation from "lib/Translation"
import { makeStyles } from "@material-ui/core/styles"
import React, { useState, Fragment } from "react"

import DialogTitle from "@material-ui/core/DialogTitle"
import Avatar from "@material-ui/core/Avatar"
import TranslateIcon from "@material-ui/icons/Translate"
import Typography from "@material-ui/core/Typography"

import DialogContent from "@material-ui/core/DialogContent"
import List from "@material-ui/core/List"
import ListItem from "@material-ui/core/ListItem"
import ListItemText from "@material-ui/core/ListItemText"
import Link from "lib/links/ReactRouterLink"

import DialogActions from "@material-ui/core/DialogActions"
import Button from "@material-ui/core/Button"

import Tooltip from "@material-ui/core/Tooltip"
import IconButton from "@material-ui/core/IconButton"

import Dialog from "@material-ui/core/Dialog"
import useSmoothClosing from "lib/TranslationDialog/useSmoothClosing"
import PropTypes from "prop-types"
const translation = new Translation({
  en: {
    currentLanguage: `Current language`,
    open: `Translation Dialog`,
    title: `Select Language`,
    close: `CLOSE`,
  },
  es: {
    currentLanguage: `Idioma actual`,
    open: `DiÃ¡logo de Idiomas`,
    title: `Seleccione un Idioma`,
    close: `CERRAR`,
  },
})
const ids = {
  dialog: `lib-tranlation_dialog`,
  title: `lib-tranlation_dialog-title`,
}
const useStyles = makeStyles(theme => ({
  title: {
    display: `flex`,
    alignItems: `center`,
    "& > *": {
      margin: theme.spacing(1),
    },
  },
  avatar: {
    backgroundColor: theme.palette.primary.main,
  },
}), { name: ids.dialog })
export default function TranslationDialog ({ getLinks }) {
  const currentTranslation = translation.get()
  const classes = useStyles()
  const [isOpen, saveIsOpen] = useState(false)
  const close = () => saveIsOpen(false)
  const children = (
    <Fragment>
      <DialogTitle id={ids.title} className={classes.title} disableTypography>
        <Avatar className={classes.avatar}><TranslateIcon /></Avatar>
        <Typography component="h2" variant="h6">{currentTranslation.title}</Typography>
      </DialogTitle>
      <DialogContent>
        <nav>
          <List>
            {isOpen && getLinks().map(({ location, otherProps }) => (
              <ListItem key={otherProps.key}>
                {location ? (
                  <ListItemText>
                    <Link id={otherProps.id} onClick={close} to={location}>
                      {otherProps.text}
                    </Link>
                  </ListItemText>
                ) : (
                  <ListItemText aria-label={currentTranslation.currentLanguage}>
                    {otherProps.text}
                  </ListItemText>
                )}
              </ListItem>
            ))}
          </List>
        </nav>
      </DialogContent>
      <DialogActions>
        <Button color="primary" aria-controls={ids.dialog} onClick={close}>
          {currentTranslation.close}
        </Button>
      </DialogActions>
    </Fragment>
  )
  return (
    <Fragment>
      <Tooltip title={currentTranslation.open}>
        <IconButton
          aria-controls={ids.dialog}
          aria-haspopup="true"
          onClick={() => saveIsOpen(true)}
        >
          <TranslateIcon />
        </IconButton>
      </Tooltip>
      <Dialog
        id={ids.dialog}
        aria-labelledby={ids.title}
        onClose={close}
        open={isOpen}
        scroll="body"
      >
        {useSmoothClosing(isOpen, children)}
      </Dialog>
    </Fragment>
  )
}
TranslationDialog.propTypes = {
  getLinks: PropTypes.func.isRequired,
}