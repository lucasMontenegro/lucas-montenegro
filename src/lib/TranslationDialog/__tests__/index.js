import React, { useState } from "react"

import DialogContent from "@material-ui/core/DialogContent"
import List from "@material-ui/core/List"
import ListItem from "@material-ui/core/ListItem"
import ListItemText from "@material-ui/core/ListItemText"
import Link from "lib/links/ReactRouterLink"

import Dialog from "@material-ui/core/Dialog"

import TranslationDialog from "lib/TranslationDialog"
import renderer from "react-test-renderer"

jest.mock(`lib/Translation`, () => ({
  __esModule: true,
  default: function () {
    this.get = () => ({})
  },
}))
jest.mock(`@material-ui/core/styles`, () => ({
  __esModule: true,
  makeStyles: () => () => ({}),
}))
jest.mock(`react`, () => {
  const react = jest.requireActual(`react`)
  return {
    ...react,
    __esModule: true,
    default: react,
    useState: jest.fn(),
  }
})
let isOpenValue
useState.mockImplementation(() => [isOpenValue, () => {}])

jest.mock(`@material-ui/core/DialogTitle`, () => ({ __esModule: true, default: () => null }))
jest.mock(`@material-ui/core/Avatar`, () => ({ __esModule: true, default: () => null }))
jest.mock(`@material-ui/icons/Translate`, () => ({ __esModule: true, default: () => null }))
jest.mock(`@material-ui/core/Typography`, () => ({ __esModule: true, default: () => null }))

function makeSimpleComponentMock (name) {
  return ({ children }) => <div className={name}>{children}</div>
}
jest.mock(`@material-ui/core/DialogContent`, () => ({ __esModule: true, default: jest.fn() }))
DialogContent.mockImplementation(makeSimpleComponentMock(`DialogContent`))
jest.mock(`@material-ui/core/List`, () => ({ __esModule: true, default: jest.fn() }))
List.mockImplementation(makeSimpleComponentMock(`List`))
jest.mock(`@material-ui/core/ListItem`, () => ({ __esModule: true, default: jest.fn() }))
ListItem.mockImplementation(makeSimpleComponentMock(`ListItem`))
jest.mock(`@material-ui/core/ListItemText`, () => ({ __esModule: true, default: jest.fn() }))
ListItemText.mockImplementation(makeSimpleComponentMock(`ListItemText`))
jest.mock(`lib/links/ReactRouterLink`, () => ({ __esModule: true, default: jest.fn() }))
Link.mockImplementation(makeSimpleComponentMock(`Link`))

jest.mock(`@material-ui/core/DialogActions`, () => ({ __esModule: true, default: () => null }))
jest.mock(`@material-ui/core/Button`, () => ({ __esModule: true, default: () => null }))

jest.mock(`@material-ui/core/Tooltip`, () => ({ __esModule: true, default: () => null }))
jest.mock(`@material-ui/core/IconButton`, () => ({ __esModule: true, default: () => null }))

jest.mock(`@material-ui/core/Dialog`, () => ({ __esModule: true, default: jest.fn() }))
Dialog.mockImplementation(makeSimpleComponentMock(`Dialog`))
jest.mock(`lib/TranslationDialog/useSmoothClosing`, () => ({
  __esModule: true,
  default: (x, y) => y,
}))
describe(`lib/TranslationDialog`, () => {
  it(`should use the right dependency versions`, () => {
    expect(jestUtils.getDependencies([
      `@material-ui/core`,
      `react`,
      `@material-ui/icons`,
      `prop-types`,
    ])).toMatchSnapshot()
  })
  describe(`TranslationDialog (isOpen false)`, () => {
    let result
    beforeAll(() => {
      isOpenValue = false
      result = renderer.create(<TranslationDialog getLinks={() => {}} />).toJSON()
    })
    it(`should not render links`, () => {
      expect(result).toMatchSnapshot()
    })
  })
  describe(`TranslationDialog (isOpen true)`, () => {
    let result
    beforeAll(() => {
      isOpenValue = true
      function getLinks () {
        return [
          { location: null, otherProps: { key: `null` } },
          { location: {}, otherProps: { key: `not null` } }
        ]
      }
      result = renderer.create(<TranslationDialog getLinks={getLinks} />).toJSON()
    })
    it(`should render links`, () => {
      expect(result).toMatchSnapshot()
    })
  })
})