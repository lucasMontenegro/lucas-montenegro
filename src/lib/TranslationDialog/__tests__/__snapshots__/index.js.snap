// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/TranslationDialog TranslationDialog (isOpen false) should not render links 1`] = `
<div
  className="Dialog"
>
  <div
    className="DialogContent"
  >
    <nav>
      <div
        className="List"
      />
    </nav>
  </div>
</div>
`;

exports[`lib/TranslationDialog TranslationDialog (isOpen true) should render links 1`] = `
<div
  className="Dialog"
>
  <div
    className="DialogContent"
  >
    <nav>
      <div
        className="List"
      >
        <div
          className="ListItem"
        >
          <div
            className="ListItemText"
          />
        </div>
        <div
          className="ListItem"
        >
          <div
            className="ListItemText"
          >
            <div
              className="Link"
            />
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;

exports[`lib/TranslationDialog should use the right dependency versions 1`] = `
Object {
  "@material-ui/core": "^4.5.0",
  "@material-ui/icons": "^4.4.3",
  "prop-types": "^15.7.2",
  "react": "^16.10.1",
}
`;
